web/src/app/components/bar-chart/bar-chart.component.ts

```diff
@@ -0,0 +1,401 @@
+ * `undefined` or `null`. The SVGSVGElement provides the access and all methods
+ * to manipulate `<svg>` element, while SVGGElement corresponds to the `g` element
+ * that the top bar chart and the bottom bar chart belong to.
+ * SVGSVGElement --> SVGGElement --> SVGElement --> Element --> Node --> EventTarget
+ * We want to separate the top bar chart and the bottom bar chart by `g` elements,
+ * so that they can be updated with different methods using dropdown menu and brush
+ * selector.
+ */
+type d3SVG = d3.Selection<SVGSVGElement, undefined, null, undefined>;
+type d3G = d3.Selection<SVGGElement, undefined, null, undefined>;
+type d3ScaleLinear = d3.ScaleLinear<number, number>;
+type d3ScaleBand = d3.ScaleBand<string>;
+
+@Component({
+  selector: 'app-bar-chart',
+  templateUrl: './bar-chart.component.html',
+  styleUrls: ['./bar-chart.component.scss']
+})
+
+export class BarChartComponent implements AfterViewInit {
+  /**
+   * Constants.
+   */
+  private static readonly ALL_PUSHES_OPTION: number = 0;
+  private static readonly DEFAULT_NUM_BARS: number = 30;
+  private static readonly NANO_TO_SECS: number = 10 ** 9;
+  private static readonly NANO_TO_MILLI: number = 10 ** 6;
+  private static readonly SECS_TO_HRS: number = 60 * 60;
+  private static readonly COLOR_LIGHT_GRAY: string = '#787878';
+  private static readonly COLOR_DARK_GRAY: string = '#373C38';
+  private static readonly COLOR_WHITE: string = '#eee';
+  private static readonly COLOR_WHITE_TRANS: string = '#ffffff00';
+  private static readonly DATE_FORMAT: string = 'yyyy-mm-dd hh:mm';
+  private static readonly DATE_LOCALE: string = 'en-US';
+  private static readonly STATE_TO_COLOR: { [index: number]: string } = {
+    1: '#eee',
+    3: '#2196f3',
+    4: '#d50000',
+    5: '#34a853',
+    6: '#d50000',
+    7: '#2196f3',
+    8: '#2196f3',
+    9: '#d50000',
+    10: '#2196f3',
+    12: '#d50000',
+    13: '#2196f3',
+    15: '#2196f3',
+    16: '#d50000'
+  };
+
+  /**
+   * Private variables.
+   *
+   * The non-null assertion operator('!') suggests that there is no
+   * issue of variables being `null` or `undefined`. Some variables are
+   * declared as any, as their types change in different functions.
+   */
+  @ViewChild('barchart') private barChartContainer!: ElementRef; // Bar chart component
+  @Input() private pushInfos!: step189_2020.IPushInfo[] | null;
+  @Input() private currentPush!: step189_2020.IPushInfo | null;
+
+  private dataAll: Item[] = [];
+  private dataComplete: Item[] = [];
+  private totalDuration: number[] = [];
+  private completeDuration: number[] = [];
+  private svg: d3SVG | undefined;
+  private focus: d3G | undefined; // Top bar chart for display.
+  private brush: d3G | undefined; // Bottom bar chart for brushing.
+  private brushSelector: any;
+  private heightFocus = 0;
+  private heightBrush = 0;
+  private width = 0;
+  private xScaleFocus: any;
+  private yScaleFocus!: d3ScaleLinear;
```

Comment:
Are there plans to get rid of `any` and `!` for these in a separate PR?

Reply:
Yes! Maybe after merging this one :)